File: OTEAServer/Controllers/CountriesController.cs
using Microsoft.AspNetCore.Mvc;
using OTEAServer.Misc;
using OTEAServer.Models;
using System.Linq.Expressions;
using System.Text.Json;
using static OTEAServer.Controllers.CitiesController;
using static OTEAServer.Controllers.CountriesController;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller for countries operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>


    [ApiController]
    [Route("Countries")]
    public class CountriesController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public CountriesController(DatabaseContext context)
        {
            _context = context;
        }

        private class CountryDto {
            public string IdCountry { get; set; }
            public string Name { get; set; }

            public string PhoneCode { get; set; }

            public string Flag { get; set; }
        }

        /// <summary>
        /// Method that obtains all the countries in language order
        /// </summary>
        /// <param name="language">App language (english, spanish, french, basque, catalan, dutch, galician, german, italian and portuguese)</param>
        /// <returns>Country list</returns>
        [HttpGet("all")]
        public IActionResult GetAll([FromQuery] string language)
        {
            try
            {
                IQueryable<Country> query = _context.Countries.AsQueryable();

                Expression<Func<Country, CountryDto>> selector = c => new CountryDto
                {
                    IdCountry = c.idCountry,
                    Name = c.nameEnglish, // Default value
                    PhoneCode = c.phone_code,
                    Flag = c.flag
                };
                switch (language)
                {
                    case "es":
                        query = query.OrderByDescending(c => c.idCountry == "ESP")
                                    .ThenByDescending(c => c.idCountry == "AND")
                                    .ThenByDescending(c => c.idCountry == "ARG")
                                    .ThenByDescending(c => c.idCountry == "BOL")
                                    .ThenByDescending(c => c.idCountry == "CHL")
                                    .ThenByDescending(c => c.idCountry == "COL")
                                    .ThenByDescending(c => c.idCountry == "CRI")
                                    .ThenByDescending(c => c.idCountry == "CUB")
                                    .ThenByDescending(c => c.idCountry == "ECU")
                                    .ThenByDescending(c => c.idCountry == "SLV")
                                    .ThenByDescending(c => c.idCountry == "GUA")
                                    .ThenByDescending(c => c.idCountry == "GNQ")
                                    .ThenByDescending(c => c.idCountry == "HND")
                                    .ThenByDescending(c => c.idCountry == "MEX")
                                    .ThenByDescending(c => c.idCountry == "NIC")
                                    .ThenByDescending(c => c.idCountry == "PAN")
                                    .ThenByDescending(c => c.idCountry == "PRY")
                                    .ThenByDescending(c => c.idCountry == "PER")
                                    .ThenByDescending(c => c.idCountry == "PRI")
                                    .ThenByDescending(c => c.idCountry == "DOM")
                                    .ThenByDescending(c => c.idCountry == "URY")
                                    .ThenByDescending(c => c.idCountry == "VEN")
                                    .ThenBy(c => c.nameSpanish);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameSpanish,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "fr":
                        query = query.OrderByDescending(c => c.idCountry == "FRA")
                                    .ThenByDescending(c => c.idCountry == "MCO")
                                    .ThenByDescending(c => c.idCountry == "BEL")
                                    .ThenByDescending(c => c.idCountry == "CHE")
                                    .ThenByDescending(c => c.idCountry == "LUX")
                                    .ThenByDescending(c => c.idCountry == "CAN")
                                    .ThenByDescending(c => c.idCountry == "MAR")
                                    .ThenByDescending(c => c.idCountry == "DZA")
                                    .ThenBy(c => c.nameFrench);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameFrench,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "eu":
                        query = query.OrderByDescending(c => c.idCountry == "ESP")
                                    .ThenByDescending(c => c.idCountry == "FRA")
                                    .ThenByDescending(c => c.idCountry == "AND")
                                    .ThenByDescending(c => c.idCountry == "ARG")
                                    .ThenByDescending(c => c.idCountry == "BOL")
                                    .ThenByDescending(c => c.idCountry == "CHL")
                                    .ThenByDescending(c => c.idCountry == "COL")
                                    .ThenByDescending(c => c.idCountry == "CRI")
                                    .ThenByDescending(c => c.idCountry == "CUB")
                                    .ThenByDescending(c => c.idCountry == "ECU")
                                    .ThenByDescending(c => c.idCountry == "SLV")
                                    .ThenByDescending(c => c.idCountry == "GUA")
                                    .ThenByDescending(c => c.idCountry == "GNQ")
                                    .ThenByDescending(c => c.idCountry == "HND")
                                    .ThenByDescending(c => c.idCountry == "MEX")
                                    .ThenByDescending(c => c.idCountry == "NIC")
                                    .ThenByDescending(c => c.idCountry == "PAN")
                                    .ThenByDescending(c => c.idCountry == "PRY")
                                    .ThenByDescending(c => c.idCountry == "PER")
                                    .ThenByDescending(c => c.idCountry == "PRI")
                                    .ThenByDescending(c => c.idCountry == "DOM")
                                    .ThenByDescending(c => c.idCountry == "URY")
                                    .ThenByDescending(c => c.idCountry == "VEN")
                                    .ThenBy(c => c.nameBasque);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameBasque,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "ca":
                        query = query.OrderByDescending(c => c.idCountry == "ESP")
                                    .ThenByDescending(c => c.idCountry == "AND")
                                    .ThenByDescending(c => c.idCountry == "ARG")
                                    .ThenByDescending(c => c.idCountry == "BOL")
                                    .ThenByDescending(c => c.idCountry == "CHL")
                                    .ThenByDescending(c => c.idCountry == "COL")
                                    .ThenByDescending(c => c.idCountry == "CRI")
                                    .ThenByDescending(c => c.idCountry == "CUB")
                                    .ThenByDescending(c => c.idCountry == "ECU")
                                    .ThenByDescending(c => c.idCountry == "SLV")
                                    .ThenByDescending(c => c.idCountry == "GUA")
                                    .ThenByDescending(c => c.idCountry == "GNQ")
                                    .ThenByDescending(c => c.idCountry == "HND")
                                    .ThenByDescending(c => c.idCountry == "MEX")
                                    .ThenByDescending(c => c.idCountry == "NIC")
                                    .ThenByDescending(c => c.idCountry == "PAN")
                                    .ThenByDescending(c => c.idCountry == "PRY")
                                    .ThenByDescending(c => c.idCountry == "PER")
                                    .ThenByDescending(c => c.idCountry == "PRI")
                                    .ThenByDescending(c => c.idCountry == "DOM")
                                    .ThenByDescending(c => c.idCountry == "URY")
                                    .ThenByDescending(c => c.idCountry == "VEN")
                                    .ThenBy(c => c.nameCatalan);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameCatalan,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "gl":
                        query = query.OrderByDescending(c => c.idCountry == "ESP")
                                    .ThenByDescending(c => c.idCountry == "AND")
                                    .ThenByDescending(c => c.idCountry == "ARG")
                                    .ThenByDescending(c => c.idCountry == "BOL")
                                    .ThenByDescending(c => c.idCountry == "CHL")
                                    .ThenByDescending(c => c.idCountry == "COL")
                                    .ThenByDescending(c => c.idCountry == "CRI")
                                    .ThenByDescending(c => c.idCountry == "CUB")
                                    .ThenByDescending(c => c.idCountry == "ECU")
                                    .ThenByDescending(c => c.idCountry == "SLV")
                                    .ThenByDescending(c => c.idCountry == "GUA")
                                    .ThenByDescending(c => c.idCountry == "GNQ")
                                    .ThenByDescending(c => c.idCountry == "HND")
                                    .ThenByDescending(c => c.idCountry == "MEX")
                                    .ThenByDescending(c => c.idCountry == "NIC")
                                    .ThenByDescending(c => c.idCountry == "PAN")
                                    .ThenByDescending(c => c.idCountry == "PRY")
                                    .ThenByDescending(c => c.idCountry == "PER")
                                    .ThenByDescending(c => c.idCountry == "PRI")
                                    .ThenByDescending(c => c.idCountry == "DOM")
                                    .ThenByDescending(c => c.idCountry == "URY")
                                    .ThenByDescending(c => c.idCountry == "VEN")
                                    .ThenBy(c => c.nameGalician);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameGalician,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "pt":
                        query = query.OrderByDescending(c => c.idCountry == "PRT")
                                    .ThenByDescending(c => c.idCountry == "BRA")
                                    .ThenByDescending(c => c.idCountry == "GNB")
                                    .ThenByDescending(c => c.idCountry == "MOZ")
                                    .ThenByDescending(c => c.idCountry == "STP")
                                    .ThenByDescending(c => c.idCountry == "TLS")
                                    .ThenBy(c => c.namePortuguese);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.namePortuguese,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "de":
                        query = query.OrderByDescending(c => c.idCountry == "DEU")
                                    .ThenByDescending(c => c.idCountry == "AUT")
                                    .ThenByDescending(c => c.idCountry == "BEL")
                                    .ThenByDescending(c => c.idCountry == "LUX")
                                    .ThenByDescending(c => c.idCountry == "LIE")
                                    .ThenByDescending(c => c.idCountry == "CHE")
                                    .ThenBy(c => c.nameGerman);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameGerman,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "it":
                        query = query.OrderByDescending(c => c.idCountry == "ITA")
                                    .ThenByDescending(c => c.idCountry == "CHE")
                                    .ThenByDescending(c => c.idCountry == "SMR")
                                    .ThenByDescending(c => c.idCountry == "VAT")
                                    .ThenBy(c => c.nameItalian);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameItalian,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    case "nl":
                        query = query.OrderByDescending(c => c.idCountry == "NLD")
                                    .ThenByDescending(c => c.idCountry == "BEL")
                                    .ThenByDescending(c => c.idCountry == "FRA")
                                    .ThenByDescending(c => c.idCountry == "ABW")
                                    .ThenByDescending(c => c.idCountry == "SUR")
                                    .ThenBy(c => c.nameDutch);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameDutch,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                    default:
                        query = query.OrderByDescending(c => c.idCountry == "USA")
                                    .ThenByDescending(c => c.idCountry == "GBR")
                                    .ThenByDescending(c => c.idCountry == "AUS")
                                    .ThenByDescending(c => c.idCountry == "CAN")
                                    .ThenByDescending(c => c.idCountry == "GIB")
                                    .ThenByDescending(c => c.idCountry == "IOT")
                                    .ThenByDescending(c => c.idCountry == "IND")
                                    .ThenByDescending(c => c.idCountry == "IRL")
                                    .ThenByDescending(c => c.idCountry == "NZF")
                                    .ThenByDescending(c => c.idCountry == "ZAF")
                                    .ThenBy(c => c.nameEnglish);
                        selector = c => new CountryDto
                        {
                            IdCountry = c.idCountry,
                            Name = c.nameEnglish,
                            PhoneCode = c.phone_code,
                            Flag = c.flag
                        };
                        break;
                }

                var countries = query.Select(selector).ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (CountryDto country in countries)
                {
                    String rg = "{\"idCountry\":\"" + country.IdCountry + "\"," +
                            "\"name\":\"" + country.Name + "\"," +
                            "\"phone_code\":\"" + country.PhoneCode + "\"," +
                            "\"flag\":\"" + country.Flag + "\"}";
                    result.Add(JsonDocument.Parse(rg));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }

        }

        /// <summary>
        /// Method that obtains the country from the database
        /// </summary>
        /// <param name="idCountry">Country identifier</param>
        /// <returns>Country if exists, null if not</returns>

        [HttpGet("get")]
        public ActionResult<Country> Get([FromQuery] string idCountry)
        {
            try
            {
                var country = _context.Countries.FirstOrDefault(c => c.idCountry == idCountry);

                if (country == null)
                    return NotFound();

                return country;
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}

File: connection/src/main/java/otea/connection/controller/CountriesController.java
package otea.connection.controller;

import com.google.gson.JsonObject;

import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.organization.data.geo.Country;
import cli.organization.data.geo.Country;
import cli.organization.data.geo.Region;
import otea.connection.ConnectionClient;
import otea.connection.api.CountriesApi;
import retrofit2.Call;
import retrofit2.Response;

/**
 * Controller for countries operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class CountriesController {

    /**Countries api used to connect to the server*/
    private static CountriesApi api;

    /**Controller instance*/
    private static CountriesController instance;

    /**Class constructor*/
    private CountriesController(){
        api= ConnectionClient.getInstance().getRetrofit().create(CountriesApi.class);
    }

    /**Number of attempts*/
    private static int numAttempts=0;

    /**
     * Method that obtains the singleton instance of the controller
     *
     * @return Controller instance
     * */
    public static CountriesController getInstance(){
        if(instance==null){
            synchronized (CountriesController.class){
                if(instance==null){
                    instance=new CountriesController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new CountriesController();
    }

    /**
     * Method that obtains the country from the database
     *
     * @param idCountry - Country identifier
     * @return Country if exists, null if not
     * */
    public static Country GetCountry(String idCountry){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<Country> callable = new Callable<Country>() {
            @Override
            public Country call() throws Exception {
                Call<Country> call = api.GetCountry(idCountry);
                Response<Country> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<Country> future = executor.submit(callable);
            Country result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that obtains all the countries in language order
     *
     * @param language - App language (english, spanish, french, basque, catalan, dutch, galician, german, italian and portuguese)
     * @return Country list
     * */
    public static List<Country> GetAll(String language){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<JsonObject>> callable = new Callable<List<JsonObject>>() {
            @Override
            public List<JsonObject> call() throws Exception {
                Call<List<JsonObject>> call = api.GetAll(language);
                Response<List<JsonObject>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<JsonObject>> future = executor.submit(callable);
            List<JsonObject> list = future.get();
            executor.shutdown();
            numAttempts=0;
            List<Country> countries=new ArrayList<>();
            for(JsonObject reg:list){
                String idCountry=reg.getAsJsonPrimitive("idCountry").getAsString();
                String name=reg.getAsJsonPrimitive("name").getAsString();
                String nameSpanish = "";
                String nameEnglish = "";
                String nameFrench = "";
                String nameBasque = "";
                String nameCatalan = "";
                String nameDutch = "";
                String nameGalician = "";
                String nameGerman = "";
                String nameItalian = "";
                String namePortuguese = "";
                if(Locale.getDefault().getLanguage().equals("es")){
                    nameSpanish=name;
                }else if(Locale.getDefault().getLanguage().equals("fr")){
                    nameFrench=name;
                }else if(Locale.getDefault().getLanguage().equals("eu")){
                    nameBasque=name;
                }else if(Locale.getDefault().getLanguage().equals("ca")){
                    nameCatalan=name;
                }else if(Locale.getDefault().getLanguage().equals("nl")){
                    nameDutch=name;
                }else if(Locale.getDefault().getLanguage().equals("gl")){
                    nameGalician=name;
                }else if(Locale.getDefault().getLanguage().equals("de")){
                    nameGerman=name;
                }else if(Locale.getDefault().getLanguage().equals("it")){
                    nameItalian=name;
                }else if(Locale.getDefault().getLanguage().equals("pt")){
                    namePortuguese=name;
                }else{
                    nameEnglish=name;
                }
                String phone_code=reg.getAsJsonPrimitive("phone_code").getAsString();
                String flag=reg.getAsJsonPrimitive("flag").getAsString();
                countries.add(new Country(idCountry,nameSpanish,nameEnglish,nameFrench,nameBasque,nameCatalan,nameDutch,nameGalician,nameGerman,nameItalian,namePortuguese,phone_code,flag));
            }
            return countries;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetAll(language);
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }
    }

}
